File: apk/app/src/main/java/com/example/momentra/ui/shell/group/shared/GroupActivityCategoryFilter.kt
package com.example.momentra.ui.shell.group.shared

import com.example.momentra.data.api.ActivityItemDto
import com.example.momentra.ui.shell.group.experience.create.ExperienceQuickAddKind
import com.example.momentra.ui.shell.group.experience.create.emoji
import com.example.momentra.ui.shell.group.living.create.LivingQuickAddKind
import com.example.momentra.ui.shell.group.living.create.emoji
import com.example.momentra.ui.shell.group.living.create.label
import com.example.momentra.ui.shell.group.purchase.create.PurchaseQuickAddKind
import com.example.momentra.ui.shell.group.purchase.create.emoji
import com.example.momentra.ui.shell.group.purchase.create.label
import com.example.momentra.ui.shell.group.wedding.create.WeddingQuickAddKind
import java.util.Locale

/** Category chips for Group “All activity”, sourced from that moment’s Quick Add hub names. */
object GroupActivityCategoryFilter {
    const val ALL_ID = "All"

    data class FilterChip(
        val id: String,
        val label: String,
        val emoji: String,
    )

    fun chips(momentTypeCode: String?): List<FilterChip> {
        val all = FilterChip(ALL_ID, "All", "📋")
        val family = groupExperienceFamilyFor(momentTypeCode)
        val hub = when {
            family.isWedding() -> WeddingQuickAddKind.entries.map {
                FilterChip(chipId(it), weddingHubLabel(it), weddingEmoji(it))
            }
            family.isThemedExperience() -> ExperienceQuickAddKind.entries.map {
                FilterChip(chipId(it), experienceHubLabel(it), it.emoji())
            }
            family.isThemedPurchase() -> PurchaseQuickAddKind.entries.map {
                FilterChip(chipId(it), it.label(), it.emoji())
            }
            family.isThemedLiving() -> LivingQuickAddKind.entries.map {
                FilterChip(chipId(it), it.label(), it.emoji())
            }
            else -> tripHubChips()
        }
        return listOf(all) + hub
    }

    fun matches(item: ActivityItemDto, chipId: String): Boolean =
        matches(item.activityCode, chipId)

    fun matches(activityCode: String, chipId: String): Boolean {
        if (chipId == ALL_ID || chipId.isBlank()) return true
        return activityBelongs(chipId, activityCode)
    }

    private fun tripHubChips(): List<FilterChip> {
        val emojiById = mapOf(
            "expense" to "💳",
            "planning" to "📋",
            "budget" to "💰",
            "booking" to "🧳",
            "poll" to "📊",
            "memory" to "📷",
            "update" to "📢",
            "contribution" to "🎁",
            "invite" to "👤",
        )
        // figmaHubTiles order already matches trip hub layout for these ids.
        return GroupActionRegistry.figmaTripHubTiles.map {
            FilterChip(it.id, it.label, emojiById[it.id] ?: "📌")
        }
    }

    private fun chipId(kind: WeddingQuickAddKind): String =
        kind.name.lowercase(Locale.US)

    private fun chipId(kind: ExperienceQuickAddKind): String =
        kind.name.lowercase(Locale.US)

    private fun chipId(kind: LivingQuickAddKind): String =
        kind.name.lowercase(Locale.US)

    private fun chipId(kind: PurchaseQuickAddKind): String = when (kind) {
        PurchaseQuickAddKind.PURCHASE_ITEM -> "purchaseItem"
        else -> kind.name.lowercase(Locale.US)
    }

    private fun weddingHubLabel(kind: WeddingQuickAddKind): String = when (kind) {
        WeddingQuickAddKind.PARTICIPANT -> "Invite"
        WeddingQuickAddKind.PLANNING -> "Planning"
        WeddingQuickAddKind.EXPENSE -> "Expense"
        WeddingQuickAddKind.BUDGET -> "Budget"
        WeddingQuickAddKind.CONTRIBUTION -> "Contribution"
        WeddingQuickAddKind.SETTLE -> "Settle"
        WeddingQuickAddKind.VENDOR -> "Vendor"
        WeddingQuickAddKind.ATTENDANCE -> "Attendance"
        WeddingQuickAddKind.UPDATE -> "Update"
        WeddingQuickAddKind.POLL -> "Poll"
        WeddingQuickAddKind.MEMORY -> "Memory"
    }

    private fun weddingEmoji(kind: WeddingQuickAddKind): String = when (kind) {
        WeddingQuickAddKind.PARTICIPANT -> "👤"
        WeddingQuickAddKind.PLANNING -> "📋"
        WeddingQuickAddKind.EXPENSE -> "💳"
        WeddingQuickAddKind.BUDGET -> "💰"
        WeddingQuickAddKind.CONTRIBUTION -> "🎁"
        WeddingQuickAddKind.SETTLE -> "⚖️"
        WeddingQuickAddKind.VENDOR -> "🏪"
        WeddingQuickAddKind.ATTENDANCE -> "✅"
        WeddingQuickAddKind.UPDATE -> "📢"
        WeddingQuickAddKind.POLL -> "📊"
        WeddingQuickAddKind.MEMORY -> "📷"
    }

    private fun experienceHubLabel(kind: ExperienceQuickAddKind): String = when (kind) {
        ExperienceQuickAddKind.PARTICIPANT -> "Invite"
        ExperienceQuickAddKind.PLANNING -> "Planning"
        ExperienceQuickAddKind.EXPENSE -> "Expense"
        ExperienceQuickAddKind.BUDGET -> "Budget"
        ExperienceQuickAddKind.CONTRIBUTION -> "Contribution"
        ExperienceQuickAddKind.SETTLE -> "Settle"
        ExperienceQuickAddKind.VENDOR -> "Vendor"
        ExperienceQuickAddKind.ATTENDANCE -> "Attendance"
        ExperienceQuickAddKind.UPDATE -> "Update"
        ExperienceQuickAddKind.POLL -> "Poll"
        ExperienceQuickAddKind.MEMORY -> "Memory"
        ExperienceQuickAddKind.BOOKING -> "Booking"
    }

    private fun activityBelongs(chipId: String, activityCode: String): Boolean {
        val upper = activityCode.uppercase(Locale.US)
        return when (chipId) {
            "expense" -> upper.contains("EXPENSE")
            "contribution" -> upper.contains("CONTRIB")
            "settle" -> upper.contains("SETTLE")
            "planning" -> upper.contains("PLANNING")
            "task" -> upper.contains("TASK") && !upper.contains("PLANNING")
            "booking" -> upper.contains("BOOKING")
            "poll" -> upper.contains("POLL")
            "memory" -> upper.contains("MEMORY")
            "update" -> upper.contains("UPDATE")
            "invite", "participant", "resident", "contributor" ->
                upper.contains("MEMBER") ||
                    upper.contains("RESIDENT") ||
                    upper.contains("PARTICIPANT") ||
                    upper.contains("INVITE") ||
                    upper.contains("CONTRIBUTOR")
            "vendor" -> upper.contains("VENDOR")
            "attendance" -> upper.contains("ATTENDANCE")
            "purchaseItem", "purchase" -> upper.contains("PURCHASE")
            "delivery" -> upper.contains("DELIVERY")
            "ownership" -> upper.contains("OWNERSHIP") || upper.contains("TRANSFER")
            "asset" -> upper.contains("ASSET")
            "maintenance" -> upper.contains("MAINTENANCE")
            "rule" -> upper.contains("RULE")
            "budget" -> upper.contains("BUDGET")
            else -> false
        }
    }
}

File: momentra/momentra/Shell/GroupActive/Shared/GroupActivityCategoryFilter.swift
import Foundation

/// Category chips for Group “All activity”, sourced from that moment’s Quick Add hub names.
enum GroupActivityCategoryFilter {
    struct FilterChip: Identifiable, Equatable {
        let id: String
        let label: String
        let emoji: String
    }

    static let allId = "All"

    static func chips(for momentTypeCode: String?) -> [FilterChip] {
        let all = FilterChip(id: allId, label: "All", emoji: "📋")
        let family = GroupExperienceFamily.forTypeCode(momentTypeCode)
        let hub: [FilterChip]
        if family.isWedding {
            hub = WeddingQuickAddKind.allCases.map { kind in
                FilterChip(id: kind.rawValue, label: weddingHubLabel(kind), emoji: kind.emoji)
            }
        } else if family.isThemedExperience {
            hub = ExperienceQuickAddKind.allCases.map { kind in
                FilterChip(id: kind.rawValue, label: kind.hubLabel, emoji: kind.emoji)
            }
        } else if family.isThemedPurchase {
            hub = PurchaseQuickAddKind.allCases.map { kind in
                FilterChip(id: kind.rawValue, label: kind.label, emoji: kind.emoji)
            }
        } else if family.isThemedLiving {
            hub = LivingQuickAddKind.allCases.map { kind in
                FilterChip(id: kind.rawValue, label: kind.label, emoji: kind.emoji)
            }
        } else {
            hub = tripHubChips()
        }
        return [all] + hub
    }

    static func matches(_ item: APIClient.ActivityItemPayload, chipId: String) -> Bool {
        matches(activityCode: item.activityCode, chipId: chipId)
    }

    static func matches(activityCode: String, chipId: String) -> Bool {
        if chipId == allId || chipId.isEmpty { return true }
        return activityBelongs(to: chipId, activityCode: activityCode)
    }

    // MARK: - Private

    private static func tripHubChips() -> [FilterChip] {
        let emojiById: [String: String] = [
            "expense": "💳",
            "planning": "📋",
            "checklist": "✅",
            "budget": "💰",
            "booking": "🧳",
            "poll": "📊",
            "memory": "📷",
            "update": "📢",
            "contribution": "🎁",
            "invite": "👤",
        ]
        let labels = Dictionary(
            uniqueKeysWithValues: GroupActionRegistry.figmaHubTiles(hasActiveMoment: true)
                .map { ($0.tileId, $0.label) }
        )
        return GroupActionRegistry.tripHubTileIds.compactMap { id in
            guard let label = labels[id] else { return nil }
            return FilterChip(id: id, label: label, emoji: emojiById[id] ?? "📌")
        }
    }

    private static func weddingHubLabel(_ kind: WeddingQuickAddKind) -> String {
        switch kind {
        case .participant: return "Invite"
        case .planning: return "Planning"
        case .checklist: return "Checklist"
        case .expense: return "Expense"
        case .budget: return "Budget"
        case .contribution: return "Contribution"
        case .settle: return "Settle"
        case .vendor: return "Vendor"
        case .attendance: return "Attendance"
        case .update: return "Update"
        case .poll: return "Poll"
        case .memory: return "Memory"
        }
    }

    /// Substring match on `activityCode`, ordered so more-specific tokens win when callers iterate chips.
    private static func activityBelongs(to chipId: String, activityCode: String) -> Bool {
        let upper = activityCode.uppercased()
        switch chipId {
        case "expense":
            return upper.contains("EXPENSE")
        case "contribution":
            return upper.contains("CONTRIBUTION")
                || (upper.contains("CONTRIB") && !upper.contains("CONTRIBUTOR"))
        case "settle":
            return upper.contains("SETTLE")
        case "planning":
            return upper.contains("PLANNING")
        case "checklist":
            return upper.contains("PLANNING")
        case "task":
            return upper.contains("TASK") && !upper.contains("PLANNING")
        case "booking":
            return upper.contains("BOOKING")
        case "poll":
            return upper.contains("POLL")
        case "memory":
            return upper.contains("MEMORY")
        case "update":
            return upper.contains("UPDATE")
        case "invite", "participant", "resident", "contributor":
            return upper.contains("MEMBER")
                || upper.contains("RESIDENT")
                || upper.contains("PARTICIPANT")
                || upper.contains("INVITE")
                || upper.contains("CONTRIBUTOR")
        case "vendor":
            return upper.contains("VENDOR")
        case "attendance":
            return upper.contains("ATTENDANCE")
        case "purchaseItem", "purchase":
            return upper.contains("PURCHASE")
        case "delivery":
            return upper.contains("DELIVERY")
        case "ownership":
            return upper.contains("OWNERSHIP") || upper.contains("TRANSFER")
        case "asset":
            return upper.contains("ASSET")
        case "maintenance":
            return upper.contains("MAINTENANCE")
        case "rule":
            return upper.contains("RULE")
        case "budget":
            return upper.contains("BUDGET")
        default:
            return false
        }
    }
}
